File: leet_easy_c++/leet_160/leet_160a.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        stack<ListNode*> stkA;
        stack<ListNode*> stkB;
        if(headA == NULL || headB == NULL) return NULL;
        while(headA){
            stkA.emplace(headA);
            headA = headA->next;
        }
        while(headB){
            stkB.emplace(headB);
            headB = headB->next;
        }
        ListNode *prev = NULL;
        while(!stkA.empty() && !stkB.empty() && stkA.top() == stkB.top()){
            prev = stkA.top();
            stkA.pop();
            stkB.pop();
        }
        return prev;
    }
};

File: leet_easy_python/leet_160/leet_160a.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        if not headA or not headB: return None
        iter1 = headA
        iter2 = headB
        cnt = 0
        while cnt<3:
            if iter1 == iter2: return iter1
            iter1 = iter1.next
            iter2 = iter2.next
            if iter1 == None:
                iter1 = headB
                cnt+=1
            if iter2 == None:
                iter2 = headA
                cnt+=1        
        return None
        
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """